### PR TITLE
Improve time performance

### PR DESCRIPTION
**Pull request by winterland1989**
Updated, This patch:

+ Change `POSIXTime` representation to a record with two `Int64` field instead of previous `Integer`, which improve many IO operations.

+ Add benchmarks to easily benchmark against system `time`, e.g. the old version.

+ Cleaned up some CPPs and style to reduce further burden.

### CODE REVIEW

**Comment by winterland1989** [comment]
Note there's thunk leak here so we didn't really create any `Integer` in `getCurrentTime`, if we mark `UTCTime` 's field to strict, running time reach to 80ns,  which is still much better than previous `getPOSIXTime` based one.

**Comment by hvr** [comment]
@winterland1989 just out of curiosity (if you have time/interest to try it), how does the performance compare to e.g https://hackage.haskell.org/package/uhttpc-0.1.1.0/docs/Network-HTTP-MicroClient.html#v:getPOSIXTimeUSecs ?

**Comment by winterland1989** [comment]
Let me give it a try ; ) 

```
benchmarking time/uhttp
time                 135.3 ns   (134.0 ns .. 136.7 ns)
                     0.999 R²   (0.999 R² .. 0.999 R²)
mean                 136.0 ns   (134.9 ns .. 137.3 ns)
std dev              3.880 ns   (3.264 ns .. 4.816 ns)
variance introduced by outliers: 43% (moderately inflated)
```

It seems IEEE `Double` is no faster on my machine.

**Comment by winterland1989** [comment]
After trying to optimize `getZonedTime`, now i strongly proposed we either add 

```
data POSIXTimeRaw = POSIXTimeRaw 
    { utSeconds :: {-# UNPACK #-} !CTime
    , utNanoSeconds :: {-# UNPACK #-} !Int64
    }
```

or change `POSIXTime` directly. The reason is that if we don't have this type as bridge and we don't want convert to/from `Integer` during `getZonedTime`, then we were force to include `HsTimeConfig.h` and duplicate code in `Data.Time.Clock.POSIX`. Which is not optimal.

**Comment by hvr** [comment]
wouldn't changing `POSIXTime` directly result in conversion functions becoming partial?

**Comment by winterland1989** [comment]
It is indeed a valid concern, new data type is necessary. what about reuse code in `clock` package? The `TimeSpec` is exactly what we need here.

**Comment by winterland1989** [comment]
I have manage to add `POSIXTimeRaw` and used it to improve many `IO` operations, before:

```
benchmarking time/getCurrentTime
time                 1.052 μs   (1.041 μs .. 1.065 μs)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 1.053 μs   (1.044 μs .. 1.064 μs)
std dev              31.67 ns   (25.04 ns .. 42.80 ns)
variance introduced by outliers: 41% (moderately inflated)

benchmarking time/getPOSIXTime
time                 248.9 ns   (247.0 ns .. 251.0 ns)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 251.6 ns   (248.9 ns .. 254.4 ns)
std dev              8.611 ns   (7.456 ns .. 10.57 ns)
variance introduced by outliers: 51% (severely inflated)

benchmarking time/getZonedTime
time                 4.659 μs   (4.620 μs .. 4.694 μs)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 4.654 μs   (4.609 μs .. 4.708 μs)
std dev              164.2 ns   (132.0 ns .. 223.0 ns)
variance introduced by outliers: 45% (moderately inflated)
```

now:

```
benchmarking time/getCurrentTime
time                 85.67 ns   (84.61 ns .. 86.77 ns)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 85.39 ns   (84.61 ns .. 86.34 ns)
std dev              2.816 ns   (2.216 ns .. 3.556 ns)
variance introduced by outliers: 51% (severely inflated)

benchmarking time/getPOSIXTime
time                 129.1 ns   (128.3 ns .. 130.2 ns)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 128.5 ns   (127.4 ns .. 129.6 ns)
std dev              3.602 ns   (3.043 ns .. 4.343 ns)
variance introduced by outliers: 42% (moderately inflated)

benchmarking time/getZonedTime
time                 869.1 ns   (861.1 ns .. 877.0 ns)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 866.5 ns   (859.3 ns .. 874.5 ns)
std dev              25.94 ns   (21.97 ns .. 31.13 ns)
variance introduced by outliers: 41% (moderately inflated)
```

**Comment by winterland1989** [comment]
There're lots of CPP, code and document clean up shoud be done, i'll push more commits after reviewing ; )

**Comment by erikd** [comment]
I'm always a bit suspicious of benchmarks where the variance is more than about 20%. Are you running these benchmarks on a laptop? If so, have you tried running them on a desktop or server machine?

File: time.cabal [hunk 1]
```diff
@@ -151,3 +150,66 @@ test-suite tests
         Test.AddDays
         Test.AddDaysRef
         Test.TestUtil
+        
+benchmark bench
+    hs-source-dirs: lib, benchmark
```
Comment (AshleyYakeley) [inline time.cabal]: Do you need to rebuild all of `lib`? I think you can remove this and all the modules in lib from the `other-modules` field.

**Comment by winterland1989** [comment]
Here's the result from our server(CentOS 7.1):

before my patch:

```
benchmarking time/getCurrentTime
time                 1.831 μs   (1.824 μs .. 1.838 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 1.832 μs   (1.827 μs .. 1.848 μs)
std dev              28.91 ns   (12.49 ns .. 55.04 ns)
variance introduced by outliers: 15% (moderately inflated)

benchmarking time/getPOSIXTime
time                 329.0 ns   (328.0 ns .. 330.2 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 328.9 ns   (328.3 ns .. 330.6 ns)
std dev              3.299 ns   (1.561 ns .. 6.396 ns)

benchmarking time/getZonedTime
time                 13.94 μs   (13.89 μs .. 13.99 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 13.94 μs   (13.91 μs .. 14.03 μs)
std dev              166.7 ns   (80.97 ns .. 301.1 ns)
```

After:

```
benchmarking time/getCurrentTime
time                 81.97 ns   (81.96 ns .. 81.98 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 81.97 ns   (81.96 ns .. 81.99 ns)
std dev              44.64 ps   (30.14 ps .. 67.80 ps)

benchmarking time/getPOSIXTime
time                 154.9 ns   (154.8 ns .. 154.9 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 154.9 ns   (154.8 ns .. 155.0 ns)
std dev              215.2 ps   (79.73 ps .. 457.8 ps)

benchmarking time/getZonedTime
time                 3.837 μs   (3.828 μs .. 3.847 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 3.837 μs   (3.833 μs .. 3.843 μs)
std dev              17.58 ns   (14.03 ns .. 22.65 ns)
```

File: time.cabal (hunk 1 above)
Comment (winterland1989) [inline time.cabal]: Since time is a bundled library, i think what we really want to do here is remove `time` from build depends, then this benchmark will not interference with system `time`. But if there's simpler way to do it then i'll make a change.

Comment (AshleyYakeley) [inline time.cabal]: Can't you just do the same thing as in the `test-suite` stanzas?

Comment (winterland1989) [inline time.cabal]: Another approach i can think of is to remove all these stanze, and make a seperate benchmark cabal project in benchmark folder.

File: lib/Data/Time/Clock/POSIX.hs [hunk 2]
```diff
@@ -1,27 +1,117 @@
...
 #else
 import Data.Time.Clock.CTimeval
+import Foreign.C.Types (CLong(..))
 #endif
 
+--------------------------------------------------------------------------------
+
+data POSIXTimeRaw = POSIXTimeRaw
+    { ptSeconds ::     {-# UNPACK #-} !Int64
+    , ptNanoSeconds :: {-# UNPACK #-} !Int64
```
Comment (AshleyYakeley) [inline lib/Data/Time/Clock/POSIX.hs]: is `ptNanoSeconds` ever negative? Would `Word64` be a better type? Or even `Word32`?

Comment (winterland1989) [inline lib/Data/Time/Clock/POSIX.hs]: It's never negative, but `CTime` wrap a `Int64`, so i just keep it here.

File: lib/Data/Time/Clock/POSIX.hs [hunk 3]
```diff
@@ -1,84 +1,127 @@
...
+
+instance Ord POSIXTime where
+    rawx `compare` rawy =
+        let POSIXTime xs xn = normalizePosix rawx
+            POSIXTime ys yn = normalizePosix rawy
+            os = compare xs ys
+        in if os == EQ then xn `compare` yn else os
+
+instance NFData POSIXTime where
+    rnf a = a `seq` ()
```
Comment (AshleyYakeley) [inline lib/Data/Time/Clock/POSIX.hs]: Do we need to `seq` on the `ptSeconds` and `ptNanoSeconds` members?

Comment (winterland1989) [inline lib/Data/Time/Clock/POSIX.hs]: I don't think it's necessary since we have already unpack those fields, but please ask for more judgement.  @simonmar  @hvr

File: lib/Data/Time/Clock/POSIX.hs [hunk 4]
```diff
@@ -1,84 +1,127 @@
...
 
--- Use hi-res POSIX time
-ctimespecToPosixSeconds :: CTimespec -> POSIXTime
-ctimespecToPosixSeconds (MkCTimespec (CTime s) ns) =
-    (fromIntegral s) + (fromIntegral ns) / 1000000000
-
-getPOSIXTime = liftM ctimespecToPosixSeconds getCTimespec
+getPOSIXTime = do
+    MkCTimespec (CTime s) (CLong ns) <- getCTimespec
+    return (POSIXTime s ns)
```
Comment (AshleyYakeley) [inline lib/Data/Time/Clock/POSIX.hs]: Hmm, on systems where `time_t` is 32 bit, this won't compile, will it? `s` would need to be converted.

Comment (AshleyYakeley) [inline lib/Data/Time/Clock/POSIX.hs]: Could you verify this? From http://hackage.haskell.org/package/base-4.9.0.0/docs/src/Foreign.C.Types.html#CTime :

    -- | Haskell type representing the C @time_t@ type.
    ARITHMETIC_TYPE(CTime,HTYPE_TIME_T)

As far as I can tell, `HTYPE_TIME_T` is equivalent to `time_t`.

File: lib/Data/Time/Clock/POSIX.hs (hunk 3 above)
Comment (hvr) [inline lib/Data/Time/Clock/POSIX.hs]: Yes, if you explicitly `{-# UNPACK #-}` a field (which itself has no lazy pointers anymore), then you get NF=WHNF. 

Or put differently, the NF=WHNF property can be recursively defined:

 - primitive types such as `Int`/`Int64`/`Char` etc are NF=WHNF
 - a constructor is NF=WHNF, if it has only strict fields with types which are themselves NF=WHNF
 - a sumtype is NF=WHNF if all its constructors are NF=WHNF

that's the basic gist at least

File: lib/Data/Time/Clock/POSIX.hs [hunk 5]
```diff
@@ -1,84 +1,127 @@
...
 
--- Use hi-res POSIX time
-ctimespecToPosixSeconds :: CTimespec -> POSIXTime
-ctimespecToPosixSeconds (MkCTimespec (CTime s) ns) =
-    (fromIntegral s) + (fromIntegral ns) / 1000000000
-
-getPOSIXTime = liftM ctimespecToPosixSeconds getCTimespec
+getPOSIXTime = do
+    MkCTimespec (CTime s) (CLong ns) <- getCTimespec
+    return (POSIXTime (fromIntegral s) (fromIntegral ns))
```
Comment (AshleyYakeley) [inline lib/Data/Time/Clock/POSIX.hs]: You can probably do `fromIntegral` directly from the `CTime` and `CLong`?

File: lib/Data/Time/Clock/POSIX.hs [hunk 6]
```diff
@@ -1,84 +1,127 @@
...
+--------------------------------------------------------------------------------
 
 -- | POSIX time is the nominal time since 1970-01-01 00:00 UTC
 --
--- To convert from a 'Foreign.C.CTime' or 'System.Posix.EpochTime', use 'realToFrac'.
---
-type POSIXTime = NominalDiffTime
+data POSIXTime = POSIXTime
+    { ptSeconds ::     {-# UNPACK #-} !Int64
+    , ptNanoSeconds :: {-# UNPACK #-} !Int64
```
Comment (AshleyYakeley) [inline lib/Data/Time/Clock/POSIX.hs]: Might as well make `ptNanoSeconds` `Word32`, given that it gets converted anyway.

File: lib/Data/Time/Clock/POSIX.hs (hunk 5 above)
Comment (winterland1989) [inline lib/Data/Time/Clock/POSIX.hs]: Aye, it should be not difference in performance so i left it for you ; )

File: lib/Data/Time/Clock/POSIX.hs (hunk 6 above)
Comment (winterland1989) [inline lib/Data/Time/Clock/POSIX.hs]: It's more convenient to use `Int64` here, because `Int64` can hold  a day length in `ps`, which save a conversion in `posixToUTCTime`.

Comment (AshleyYakeley) [inline lib/Data/Time/Clock/POSIX.hs]: I'll benchmark it to see if it's any slower.

Comment (AshleyYakeley) [inline lib/Data/Time/Clock/POSIX.hs]: I believe GHC will word-align `UNPACK` fields anyway.

File: lib/Data/Time/Clock/POSIX.hs (hunk 5 above)
Comment (AshleyYakeley) [inline lib/Data/Time/Clock/POSIX.hs]: I would prefer everything done in one PR rather than having to clean up afterwards...

Comment (winterland1989) [inline lib/Data/Time/Clock/POSIX.hs]: It turns out `CTime` doesn't have an `Integral` instance, and `CLong` 's `fromIntegral` rely on rewrite rules to do our translation here, so It's better to keep it as it is IMHO.

**Comment by AshleyYakeley** [comment]
This is close. However, it contains a number of different changes combined into one PR. I might push some of those changes separately just to make the others easier to examine.

File: lib/include/HsConfigure.h [hunk 7]
```diff
@@ -1,7 +0,0 @@
-#define HAS_DataPico 0
-#if LANGUAGE_Rank2Types
-#if MIN_VERSION_base(4,2,0)
-#undef HAS_DataPico
-#define HAS_DataPico 1
-#endif
-#endif
```
Comment (AshleyYakeley) [inline lib/include/HsConfigure.h]: @hvr, does the time library need to support compilers other than GHC?

**Comment by winterland1989** [comment]
Thanks, I'll push other patches after you finished this one.

**Comment by winterland1989** [comment]
Btw, I have a commit add benchmark based on `PackageImports`, i add a new cabal file in `benchmark` which import both current modules and system `time`, it makes benchmark much easier, do you want me to push it now or wait for you finish current patches?

**Comment by winterland1989** [comment]
An confusing thing just happened, at after mark all kinds of strict field, `formatTime` became slower than original one:

```
benchmarking format/old formatTime
time                 11.80 μs   (11.67 μs .. 11.91 μs)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 11.82 μs   (11.71 μs .. 11.96 μs)
std dev              404.8 ns   (328.8 ns .. 511.2 ns)
variance introduced by outliers: 41% (moderately inflated)

benchmarking format/now formatTime
time                 15.07 μs   (14.91 μs .. 15.28 μs)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 15.15 μs   (14.99 μs .. 15.30 μs)
std dev              532.1 ns   (446.9 ns .. 666.5 ns)
```

+ If i remove strict and unpack pragmas from `LocalTime/TimeOfDay/UTCTime`, `formatTime` become as fast as original one again.

+ If i add strict bang to `UTCTime`, `formatTime` will be a litter faster than original one. 

Is there something tricky happened when ghc try to optimize `formatTime`? Please shed some light.

**Comment by winterland1989** [comment]
FYI, my benchmark suggest to remove the strcit bang in `LocalTime/TimeOfDay/UTCTime` and keep `TimeZone/POSIXTime` 's strcit and unpack mark. It seems strict bang in these high level type prevent some rules/optimization to fire.

File: time.cabal [hunk 8]
```diff
@@ -150,66 +150,3 @@ test-suite tests
         Test.AddDays
         Test.AddDaysRef
         Test.TestUtil
-        
-benchmark bench
```
Comment (AshleyYakeley) [inline time.cabal]: Why did you remove this?

**Comment by winterland1989** [comment]
Pushed new bench code, you can toggle between #ff9488f and #fd3053b to see how strictness affect `formatTime`.

BTW, i'm working on a rewrite of `formatTime/FormatTime`, early result is very promising, i'll make new PR after this one get merged.

**Comment by winterland1989** [comment]
I decided that the formatting work will be continued in a separated package. Is there anything i can do on current PR to push it forward?

**Comment by AshleyYakeley** [comment]
There are a number of additional changes that make it awkward, as well as formatting changes and so on. It's not clear to me that I can remove `HsConfigure.h`, for example. As it is, I'm going to have to cherry-pick it.

**Comment by winterland1989** [comment]
OK, i can rearrange and clean these commits, and send in another PR if you want me to ; )

**Comment by AshleyYakeley** [comment]
Let me push what I've done so far, first.

**Comment by AshleyYakeley** [comment]
I've merged some of this, with various changes.

**Comment by AshleyYakeley** [comment]
The rationale for which fields should be lazy, strict, or unpacked, is not clear to me.

**Comment by winterland1989** [comment]
At first i thought mark those fields strict could avoid thunk leak during type conversions, so i marked almost every time type. But further benchmarks showed doing this had a great effect on formatting speed which i can't find a explanation, so i removed those strict marker which affect formatting.

I suspect there must be something wrong going on with current `UTCTime` 's formatting code, which is clocked at 15 us, much larger than `ZonedTime` and `LocalTime`(less than 3 us), but i can't see any problem from the code. 

BTW. I'm trying a much clearer format design [here](https://github.com/winterland1989/time/blob/new-format-cps/lib/Data/Time/Format/Part.hs), This new design won't interfere with stict field at all, from benchmark `UTCTime` 's formatting is improved to less than 5 us, but for other types(`LocalTime/ZonedTime`) there're some regressions(10%~20%). It's more than welcome if you can give it a review ; )

**Comment by AshleyYakeley** [comment]
Does it pass the formatting test? The test suite is pretty comprehensive in comparing the formatting behaviour with `strftime` in libc.

**Comment by winterland1989** [comment]
I'll try to give it a shot, but in fact the old formatting code is already failing on my Mac, last time i looked into it, it seems `strftime` on MacOS doesn't various modifier like `time` do, I'll enable tests on travis CI to see if linux box is getting any better.

**Comment by winterland1989** [comment]
I have managed to run tests with my new format branch on a ubuntu machine, it passed all tests:

```
Running 2 test suites...
Test suite ShowDefaultTZAbbreviations: RUNNING...
Test suite ShowDefaultTZAbbreviations: PASS
Test suite logged to: dist/test/time-1.7-ShowDefaultTZAbbreviations.log
Test suite tests: RUNNING...
Test suite tests: PASS
Test suite logged to: dist/test/time-1.7-tests.log
2 of 2 test suites (2 of 2 test cases) passed.
```

If you need these log file, i could send them to you. The reason of failures on MacOS is that we are following GCC's `strftime` rather than unix specifiction, for example `%Y` is default to zeros padded on unix specifiction but with no padding on GCC one.

**Comment by winterland1989** [comment]
Close since most of the change is re-implemented.
